fix(dashboard): enforce locale-specific plural forms in translations

radeknapora opened the pull request:
## Summary

**What** — What changes are introduced in this PR?

This PR adds locale-aware pluralization validation to the i18n system, ensuring each language uses only its grammatically correct plural forms (`zero`, `one`, `two`, `few`, `many`, `other`) according to CLDR/i18next rules.

**Why** — Why are these changes relevant or necessary?  

Different languages require different plural forms (English needs 2, Polish needs 4, Arabic needs 6, Chinese needs 1). Without validation, translations can have:
- Missing required forms (e.g., Polish without `few`/`many`)
- Unnecessary dummy keys (e.g., English with unused `few`/`many`)
- Incorrect UI text like "2 items" instead of "2 przedmioty" in Polish

**How** — How have these changes been implemented?

1. **i18n Tooling:**
   - Updated `scripts/i18n/generate-schema.js` to detect plural base keys and auto-include all 6 CLDR forms (`zero|one|two|few|many|other`) as optional
   - Enhanced `scripts/i18n/validate-translation.js` to validate translations against both the schema and per-locale plural requirements using `plural-config.json`
   - Added `src/i18n/plural-config.json` mapping locales to their required plural forms
2. **Schema:**
   - Regenerated `translations/$schema.json` to allow all plural forms for relevant keys and adjusted `required` lists (removing hardcoded `one/other` requirements)
3. **Tests:**
   - Revised `translations/__tests__/validate-translations.spec.ts` to check required vs extra keys and assert plural forms for `en` match `plural-config`
4. **Translations:**
   - Updated `translations/en.json` to include explicit plural key `wrongQuantity_one` under `orders.fulfillment.error`

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

- Try adding incorrect plural forms to any language file (e.g., add `items_few` to `en.json`) — validation should fail
- Try removing required plural forms from Polish translations — validation should fail
- Verify schema generation: should create all 6 plural forms as optional

## Examples

### plural-config.json structure
```json
{
  "en": ["one", "other"],
  "pl": ["one", "few", "many", "other"],
  ...
}
```

### Validation behavior
```json
//  Valid English translation
{
  "items_one": "{{count}} item",
  "items_other": "{{count}} items"
}

//  Invalid - English doesn't use 'few'
{
  "items_one": "{{count}} item",
  "items_few": "{{count}} items",  // Validation fails: extra key
  "items_other": "{{count}} items"
}

// Valid Polish translation
{
  "items_one": "{{count}} przedmiot",
  "items_few": "{{count}} przedmioty",
  "items_many": "{{count}} przedmiotów",
  "items_other": "{{count}} przedmiotu"
}

// Invalid - Polish requires 'few' and 'many'
{
  "items_one": "{{count}} przedmiot",
  "items_other": "{{count}} przedmiotów"  // Validation fails: missing 'few' and 'many'
}
```

## Checklist

Please ensure the following before requesting a review:
- [x] I have added a changeset for this PR
  - Every non-breaking change should be marked as a patch
  - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant tests
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

## Additional Context

**References:**
- [Unicode CLDR Plural Rules](https://cldr.unicode.org/index/cldr-spec/plural-rules)
- [i18next Pluralization](https://www.i18next.com/translation-function/plurals)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds locale-aware pluralization validation and schema support, with tests and minor English translation adjustments.
> 
> - **i18n Tooling**:
>   - Update `scripts/i18n/generate-schema.js` to detect plural base keys and expose all forms (`_zero|_one|_two|_few|_many|_other`).
>   - Enhance `scripts/i18n/validate-translation.js` to validate per-locale plural requirements via `src/i18n/plural-config.json` and Ajv schema.
> - **Schema**:
>   - Regenerate `src/i18n/translations/$schema.json` to include all plural forms for detected keys and relax hardcoded `required` for plural pairs.
> - **Tests**:
>   - Revise `src/i18n/translations/__tests__/validate-translations.spec.ts` to assert required vs extra keys and enforce English plural forms per `plural-config.json`.
> - **Translations**:
>   - Add `orders.fulfillment.error.wrongQuantity_one` to `en.json`.
> - **Changeset**:
>   - Add patch entry describing plural-form enforcement.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b65eb0732bd18cc00a4414654179228384ed10e6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

radeknapora committed:
fix(schema): preserve plural forms (few/many) in Polish and similar languages

Extended the i18n translation schema to support 'few' and 'many' pluralization forms alongside existing 'one' and 'other' forms. This change improves support for languages with more complex pluralization rules.

willbouch reviewed (approved):
Thank you for the contribution! I did not know that about slavic languages so definitely learned something today 😅. For the "required or not" question, I do not believe we have something in place to add conditions but I think it could be a good idea. So, I will leave the ball into your court if that is something you want to work on. You can tag me as a reviewer of that PR if you decide to do it! For now though, it is fine to push it like this so we at least get rid of the stripping out issue

willbouch committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora commented:
Sounds good 👍

Then let’s keep it like this for now to avoid the stripping issue. I’ll look into creating a follow-up PR that introduces conditional requirements for plural forms when I get a chance. I’ll make sure to tag you as a reviewer on that one.

willbouch commented:
> Sounds good 👍
> 
> Then let’s keep it like this for now to avoid the stripping issue. I’ll look into creating a follow-up PR that introduces conditional requirements for plural forms when I get a chance. I’ll make sure to tag you as a reviewer on that one.

if you can also add a changeset before merging! Sorry I forgot

radeknapora committed:
Add changeset

radeknapora commented:
Sure, I’ve already added a changeset 👍 thanks for the reminder!

willbouch commented on .changeset/few-wolves-count.md:
```diff
@@ -0,0 +1,5 @@
+"@medusajs/dashboard": patch
+---
+
+Fix schema to preserve plural forms (`few` and `many`) for Polish translations and similar languages.
```
Usually, we try to have it like the PR description you have, So fix/feat/chore(packages impacted): description

radeknapora committed:
Fix changeset

radeknapora commented:
Updated the changeset to follow the convention, is it okay now?

adrien2p reviewed (commented):
thank you for your contribution 💪  @olivermrbl how do we want to handle this one?

willbouch commented:
Bump @olivermrbl

willbouch committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

willbouch commented:
@radeknapora There seems to still be issues with unit tests at the moment. Any chance you can take a look at this?

willbouch committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

willbouch committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

willbouch commented:
Hi @radeknapora :) Are you able to take a look at the unit tests for this PR?

radeknapora commented:
Hi @willbouch sorry I missed your earlier message. I'm currently on holiday and don’t have time to look into the unit tests right now. I’ll take a look at this PR once I’m back.

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
WIP: Proposed solution for CLDR-compliant i18n pluralization

PROBLEM WE'RE TRYING TO SOLVE

Current i18n validation requires ALL languages to have identical keys. This breaks
when different languages need different plural forms:

- English needs: _one, _other (2 forms)
- Polish needs: _one, _few, _many, _other (4 forms)
- Arabic needs: _zero, _one, _two, _few, _many, _other (6 forms)

When Polish adds _few/_many, tests fail because English doesn't have them.
This violates i18next/CLDR best practices.

PROPOSED SOLUTION (WIP)

This is a working but incomplete implementation of language-aware validation:

What Works:
- Schema merges keys from ALL translation files (not just English)
- Plural forms marked as optional in schema
- Each language validated against its CLDR plural requirements
- Tests pass with warnings for incomplete translations

What's Missing:
- NO IDE WARNINGS - VS Code won't warn about missing _few in Polish files
  because they're optional in the shared schema
- Could generate per-language schemas ($schema-pl.json, etc.) but requires
  maintaining 29+ schema files

CHANGES

New:
- scripts/i18n/plural-rules.js - CLDR plural rules for all 29 languages

Modified:
- scripts/i18n/generate-schema.js - Merges all languages, marks plurals optional
- src/i18n/translations/__tests__/validate-translations.spec.ts - Language-aware validation
- src/i18n/translations/$schema.json - Regenerated

QUESTIONS FOR TEAM

1. Is this approach acceptable? (Tests pass but with warnings)
2. Should we generate per-language schemas for IDE support?
3. Better ideas for validating language-specific plural forms?

Note: This unblocks Polish translations but needs team discussion on the
best long-term solution.

willbouch commented:
Hey @radeknapora and thanks for the detailed explanation. My two-cent on this at first glance is that we should not be maintaining a schema for each language, as you said, it would become painful. I think I would personally be fine if we lose IDE validation for those plural forms as long as they end up flagged in the tests when we open the PR In other words, I believe we need the tests to be failing... My goal here is just to ensure that they are not forgotten. That being said, I am not very aware of pluralization rules for other languages, but can languages be grouped into categories? We already have a config file called `languages.ts`, so maybe we could add a config in there that specifies what is the "pluralization category" and enforce that in the tests so that English does not fail if there are missing `items_few` for example? What do you think?

radeknapora commented:
Hey @willbouch!
Great idea! You're totally right that tests need to fail, not just warn.
Let me think through the best way to integrate the pluralization categories with `languages.ts` and I'll come back with a concrete proposal for the implementation.

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
feat(i18n): implement CLDR-compliant plural validation with language-specific rules

- Add plural-config.json with CLDR plural forms for all 29 languages
- Generate schema from en.json with all plural forms as optional
- Implement recursive plural validation in validate-translation.js
- Change validate-translations for en.json plural correctness
- Tests now FAIL (not warn) on missing/extra plural forms

Breaking: Tests will fail if plural forms don't match language requirements

Fixes: Unblocks Polish and other non-English translations

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

willbouch commented on packages/admin/dashboard/scripts/i18n/validate-translation.js:
```diff
@@ -1,47 +1,135 @@
+const requiredForms = pluralConfig[langCode]
+
+if (!requiredForms) {
+  console.error(`Language "${langCode}" not found in plural-config.json`)
```
Maybe there should be a default? Most of the languages use what kind of system?

radeknapora replied:
Most languages use the same plural system as English (`one`, `other`). I intentionally didn’t include a default, though, because I want to ensure that whenever someone adds a new language, they also explicitly define its plural rules in `plural-config.json`.

This prevents silent errors for example, adding a language that appears to work, but actually requires more plural forms.

The idea is that after this PR is merged, we’ll add a note to the documentation explaining that every new locale must include its plural configuration. That way we keep consistency and avoid breaking pluralization unintentionally.

willbouch commented on packages/admin/dashboard/src/i18n/plural-config.json:
```diff
@@ -0,0 +1,32 @@
+  "no": ["one", "other"],
+  "pl": ["one", "few", "many", "other"],
+  "pt": ["one", "other"],
+  "pt-BR": ["one", "other"],
```
pt-PT will be merged before this one. Could you take care of it?

willbouch replied:
https://github.com/medusajs/medusa/pull/13935

radeknapora replied:
Done :)

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
Update plural config and translation schema

willbouch commented:
@fPolic @olivermrbl @adrien2p @carlos-r-l-rodrigues

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
Update changeset

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora commented:
Bump @willbouch @olivermrbl @NicolasGorga

shahednasser commented on packages/admin/dashboard/scripts/i18n/generate-schema.js:
```diff
@@ -20,12 +31,28 @@ function generateSchemaFromObject(obj) {
 
   const properties = {}
   const required = []
+  const localPluralBaseKeys = new Set()
```
```suggestion
  const localePluralBaseKeys = new Set()
```

shahednasser commented on packages/admin/dashboard/scripts/i18n/generate-schema.js:
```diff
@@ -20,12 +31,28 @@ function generateSchemaFromObject(obj) {
+
     properties[key] = generateSchemaFromObject(value)
     required.push(key)
   })
```
q: why aren't we doing this in a single loop?

shahednasser commented on packages/admin/dashboard/scripts/i18n/generate-schema.js:
```diff
@@ -20,12 +31,28 @@ function generateSchemaFromObject(obj) {
+    ALL_PLURAL_FORMS.forEach((form) => {
+      properties[`${baseKey}_${form}`] = { type: "string" }
+    })
+  })
```
suggestion: similarly, I think this can be part of the same loop.

shahednasser commented on packages/admin/dashboard/src/i18n/plural-config.json:
```diff
@@ -0,0 +1,33 @@
+  "uk": ["one", "few", "many", "other"],
+  "vi": ["other"],
+  "zh": ["other"]
+}
```
Add newly added languages so that the validation doesn't fail.

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
resolve review

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl

radeknapora committed:
Merge branch 'develop' into fix(dashboard)/pluralization-schema-pl
